pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,352 @@
+                    name[loc] = pl_pauli_to_idx[p]
+            else:
+                name[op.wires[0]] = pl_pauli_to_idx[op.name]
+            hamiltonian_basis_names.append(tuple(name))
+        self.stepsize = stepsize
+
+        @batch_transform
+        def algebra_commutator(tape, observables):
+            """Calculate the Lie gradient with the parameter shift rule (see `get_omegas`)"""
+            tapes = []
+            for obs in observables:
+                tapes_plus = [
+                    JacobianTape(p + "_p") for p in self.lie_algebra_basis_names
+                ]
+                tapes_min = [
+                    JacobianTape(p + "_m") for p in self.lie_algebra_basis_names
+                ]
+
+                # loop through all operations on the input tape
+                for op in tape.operations:
+                    for t in tapes_plus + tapes_min:
+                        with t:
+                            qml.apply(op)
+                for i, t in enumerate(tapes_plus):
+                    with t:
+                        qml.PauliRot(
+                            np.pi / 2,
+                            self.lie_algebra_basis_names[i],
+                            wires=list(range(self.nqubits)),
+                        )
+                        for o in obs:
+                            qml.expval(o)
+                for i, t in enumerate(tapes_min):
+                    with t:
+                        qml.PauliRot(
+                            -np.pi / 2,
+                            self.lie_algebra_basis_names[i],
+                            wires=list(range(self.nqubits)),
+                        )
+                        for o in obs:
+                            qml.expval(o)
+                tapes.append((tapes_plus, tapes_min))
+            return tapes, None
+
+        self.algebra_commutator = algebra_commutator
+
+    def step_and_cost(
+        self,
+        *args,
+        **kwargs,
+    ):
+        r"""Update the circuit with one step of the optimizer and return the corresponding
+        objective function value prior to the step.
+
+        Args:
+            *args:
+            **kwargs:
+
+        Returns:
+
+        """
+        omegas = self.get_omegas()
+        non_zero_lie_algebra_elements = []
+        non_zero_omegas = []
+        for i, element in enumerate(omegas):
+            if not np.isclose(element, 0):
+                non_zero_lie_algebra_elements.append(
+                    tuple(
+                        self.pauli_str_to_int[p]
+                        for p in self.lie_algebra_basis_names[i]
+                    )
+                )
+                non_zero_omegas.append(-omegas[i])
+        lie_gradient = qml.Hamiltonian(
+            non_zero_omegas,
+            [self.get_pauli_op(ps) for ps in non_zero_lie_algebra_elements],
+        )
+        temp_circuit = copy.copy(self.circuit)
+
+        @append_time_evolution(lie_gradient, self.stepsize)
```

Comment:
@therooler you can apply the tape transform directly to the tape if you want, to avoid this additional function wrapper?

e.g.,

```python
new_circuit = append_time_evolution(circuit.func)
self.new_qnode = qml.QNode(new_circuit, circuit.device)
```